app/models/user.rb

```diff
@@ -1,8 +1,8 @@
 class User < ApplicationRecord
-  has_secure_password
-
   before_save :downcase_email
-  before_create :create_activation_digest
+
+  devise :database_authenticatable, :registerable,
+         :recoverable, :rememberable, :validatable, :confirmable
```

Comment:
Các module này có tác dụng gì em? Cái rememberable khác gì so vs remember me thông thường

Reply:
Em xin phép trả lời ạ
1. tác dụng của các module trong devise
+ database_authenticatable: tự động mã hoá và lưu mật khẩu của user và cho phép user đăng nhập ( = sử dụng has_secure_password và BCrypt trong việc tạo, mã hóa, lưu mật khẩu của người dùng sau khi đăng ký và kiểm tra mật khẩu người dùng sau khi đăng nhập)

+ registerable: module dành cho user mới đăng ký. Cái này em cảm thấy nó hơi trùng với module trên, nhưng có 1 chút bổ sung nhỏ ở module này so với module trên đó là nó thêm route (đến trang sign up) vào routes. Nếu bỏ module này đi thì khi chạy rails routes sẽ ko còn có các route liên quan đến devise/registration#...

+ recoverable: Phục vụ cho tính năng khôi phục mật khẩu. Cho phép người dùng đặt lại mật khẩu thông qua link được gửi trong email.

+ rememberable: phiên đăng nhập sẽ được ghi nhớ và user ko cần đăng nhập lại. khi chọn remember thì đăng nhập xong, cookie sẽ có thêm key remember_user_token, và db có thêm giá trị ở trường remember_created_at. khi hết phiên đăng nhập (đăng xuất/hết hạn cookie) thì giá trị trường này set null.

+ validatable: kiểm tra, validate thông tin username và mật khẩu. Mặc định devise kiểm tra email thông qua email's regex của gem, và check độ dài của password có từ 6-128 kí tự hay không. Ngoài ra thì nếu user model có validates :email và password thì việc thêm :validatable có thể không cần thiết.

+ confirmable: người dùng mới cần click vào 1 link confirm được gửi qua email để kích hoạt tài khoản. Tránh trường hợp đăng ký bằng fake mail. Module giúp tạo 1 confirmation_token khi tài khoản mới được tạo và set null sau khi kích hoạt bằng cách click vào link được gửi trong email
Ngoài ra còn có:
+ timeoutable: thu hồi phiên (sign out) của người dùng nếu người dùng không làm gì trong  khoảng time nào đó
+ lockable: khóa hoặc mở khóa người dùng. User bị khóa không thể đăng nhập. Lý do khóa có thể là user cố gắng đăng nhập nhưng sai pass nhiều lần, user ko còn dùng hệ thống nữa.
+ trackable: truy vết người dùng, các thông tin theo dõi như: user đã đăng nhập bao nhiêu lần, đăng nhập mới nhất lúc nào, ...
+ omniauthable: cho phép người dùng đăng nhập bằng cách sử dụng session từ Facebook, Google,...

2, So sánh rememberable với cách thủ công
- Thủ công: tự tạo remember_digest (saved to db), cookie thì giữ remember_token, token được mã hóa và so sánh với digest mỗi khi user quay lại trình duyệt.ư
- Divise: Trong db chỉ lưu remember_created_at (thời gian bắt đầu ghi nhớ phiên). giá trị của cookie lưu trên trình duyệt sau sau khi được giải mã là [user_id, salt_of_password_hash_digest, timestamp]
++ timestamp: thời gian cookie được tạo
++ salt_of_password_hash_digest: thêm 1 chuỗi dữ liệu ngẫu nhiên vào password và hash.
=> điểm khác biệt là cách tạo ra remember_token